### PR TITLE
Change leak check and suspicious auto trait lint warning messages

### LINKED ISSUE #93367: Tracking Issue for sound builtin auto trait impls
This is the **summary issue** for the `SUSPICIOUS_AUTO_TRAIT_IMPLS` future-compatibility warning. The goal of this page is to describe why this lint was added and how you can fix code that is affected by it. It also provides a place to ask questions
or register a complaint if you feel the change should not be made. For more information on the policy around future-compatibility warnings, see our [breaking change policy guidelines](https://github.com/rust-lang/rfcs/blob/master/text/1122-language-semver.md).

### What is the warning for?

Rust has some special traits, called either "auto traits" or  "traits with a default impl", which are automatically implemented for your types under certain conditions. The most notable examples for this are the traits `Send` and `Sync`. A type implements these trait automatically if all its fields also implement this trait. This means that types like `struct MyData { data: Vec<String> }` automatically implement `Send` without the user having to think about it.

There are some cases where the builtin impls are not enough. For traits like `Send`, `Sync` and `Unpin` these builtin impls can [even be unsound](https://github.com/rust-lang/rust/issues/41622) when **implementing** abstractions using unsafe code. Because of this, it is allowed to also manually implement these traits, disabling the builtin impl for your type in the process.

The way we currently disable builtin impls has some quite subtle bugs, causing unsoundness: https://github.com/rust-lang/rust/issues/84857. This unsoundness is luckily quite difficult to exploit, but can be triggered under the following conditions:
- There exists an explicit specialized impl of an auto trait which has a stronger `where`-clauses than the type definition. We consider an impl to be "specialized" here if the generic arguments of the `Self` type of the impl are not simply unique generic parameters.  `impl Trait for X` is specialized if `X` is `Vec<i32>`, `HashMap<String, V>` or even `HashMap<T, T>` and not specialized for `Vec<T>` or `HashMap<K, V>`. 
- The builtin impl for the self type of that explicit impl applies for at least one concrete type rejected by the `where`-clauses of the explicit impl

[EXAMPLE](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=06a3450783bf408d6639aae27fe4e42f)

To fix this unsoundness, we intend to simplify the rules for builtin auto trait implementations. Right now, we only discard the builtin implementation if an explicit impl applies to the given type while ignoring `where`-clauses of that explicit impl. This will be changed to just always discard the builtin impl once at least one explicit impl for that type exists.

As this can cause types to stop implementing auto traits if a specialized impl for that type exists, we now warn again specialized impls if builtin impls could apply for your type. If the builtin impl would definitely not apply for your type, this change will not affect you, and we therefore try to not emit a warning in these cases.

### When will this change take effect?

After a [crater run](https://github.com/rust-lang/crater#crater) detected that our intended change to make builtin auto trait impls sound causes [some breakage](https://github.com/rust-lang/rust/pull/85048#issuecomment-1017877458) in the wider ecosystem.

We are currently searching for ways to lessen the impact by modifying the implementation to be more similar to the current situation, while still being sound. If that is not possible, we will introduce this change after the future compatibility lint has been on stable for long enough to reduce the impact of landing this change and if there are existing workarounds for breakage caused by the new rules.

Some of the cases which broke need either marker traits (#29864) or negative impls (#68318) affecting coherence to get back the old behavior. Both of these are currently not too far away from stabilization, so this will probably remain as a future compatibility warning until one of these ends up getting stabilized.

### How to fix this?

This depends on why the explicit impl has been added in the first place. If you do not want the builtin impl to apply, you may add a field to your type which definitely does not implement the given auto trait. You may also be able to remove the explicit impl entirely, if that case is already currently covered by the builtin impl. For anything else, please report your case here so that we know about the issue and are able to extend this section.

**If your code triggered this lint, please tell us about it in this issue so we may take it into consideration when deciding on how to proceed.**

### PR DESCRIPTION
**Pull request by spastorino**
The leak check lint message "this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!" is misleading as some cases may not be phased out and could end being accepted. This is under discussion still.

The suspicious auto trait lint the change in behavior already happened, so the new message is probably more accurate.

r? @lcnr 

Closes #93367

### CODE REVIEW

**Comment by Jules-Bertholet** [comment]
I think the "a future release" wording from the original (vs "the future") is valuable. Perhaps even go further, "a future Rust release"?

**Comment by Jules-Bertholet** [comment]
Another special case the lint should ideally have, which is [recommended by UTS 55](https://www.unicode.org/reports/tr55/#General-Security-Profile), is for ZWJ/ZWNJ. But that one is much harder to implement.

**Comment by spastorino** [comment]
> I think the "a future release" wording from the original (vs ambiguous "the future") is valuable. Perhaps even go further, "a future Rust release"?

Yes, what you suggest seems also good to me. Would leave a decision up to @lcnr when they do a review.

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 1]
```diff
@@ -1501,7 +1501,7 @@ declare_lint! {
     Warn,
     "distinct impls distinguished only by the leak-check code",
     @future_incompatible = FutureIncompatibleInfo {
-        reason: FutureIncompatibilityReason::FutureReleaseErrorDontReportInDeps,
+        reason: FutureIncompatibilityReason::Custom("the behavior is likely to change in the future"),
```
Comment (lcnr) [inline compiler/rustc_lint_defs/src/builtin.rs:1504]:
```suggestion
        reason: FutureIncompatibilityReason::Custom("the behavior may change in a future release"),
```

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 2]
```diff
@@ -4098,7 +4098,7 @@ declare_lint! {
     Warn,
     "the rules governing auto traits have recently changed resulting in potential breakage",
     @future_incompatible = FutureIncompatibleInfo {
```
[the remark this answers is not in the document]
Comment (spastorino) [reply compiler/rustc_lint_defs/src/builtin.rs:4063]: Before proceeding, what do we want to do here?

This ...

```diff
diff --git a/compiler/rustc_lint_defs/src/builtin.rs b/compiler/rustc_lint_defs/src/builtin.rs
index 8a3837027d6..65ac76fd7e7 100644
--- a/compiler/rustc_lint_defs/src/builtin.rs
+++ b/compiler/rustc_lint_defs/src/builtin.rs
@@ -4060,10 +4060,6 @@
     pub SUSPICIOUS_AUTO_TRAIT_IMPLS,
     Warn,
     "the rules governing auto traits have recently changed resulting in potential breakage",
-    @future_incompatible = FutureIncompatibleInfo {
-        reason: FutureIncompatibilityReason::Custom("has potentially surprising behavior"),
-        reference: "issue #93367 <https://github.com/rust-lang/rust/issues/93367>",
-    };
 }

 declare_lint! {
 ```

Or maybe that + also adding more information/changing "the rules governing auto traits have recently changed resulting in potential breakage"?.

[the remark this answers is not in the document]
Comment (lcnr) [reply compiler/rustc_lint_defs/src/builtin.rs:4063]: :shrug: we would also have to update #93367.

I kind of tend towards removing that lint completely. While there is some use in telling users "hey, this code may not do what you expect", I feel unsure about how much effort we should spend on this.

I guess for the lint, we'd change it to

```
warning: cross-crate traits with a default impl, like `Send`, should not be specialized
  --> $DIR/issue-106755.rs:17:1
   |
LL | impl !Send for TestType<i32> {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = warning: this disables the builtin implementation for all instances of `TestType<T>`
   = note: for more information, see TODO
```

where TODO is either an updated #93367 or a section in the reference

[the remark this answers is not in the document]
Comment (spastorino) [reply compiler/rustc_lint_defs/src/builtin.rs]: Ohh hold on, pushed an old change ... going to completely remove it in a bit.

[the remark this answers is not in the document]
Comment (spastorino) [reply compiler/rustc_lint_defs/src/builtin.rs]: I think we still want this https://github.com/spastorino/rust/blob/4859591d909a01c3c213fe8ddb59fe872475ba74/compiler/rustc_hir_analysis/src/coherence/orphan.rs#L499-L528 in that case, what lint should we use there?.

**Comment by lcnr** [comment]
removing the `suspicious_auto_trait_impls` lint should just accept the code, not always error in these cases.

File: compiler/rustc_hir_analysis/src/coherence/orphan.rs [hunk 3]
```diff
@@ -26,12 +18,7 @@ pub(crate) fn orphan_check_impl(
     let trait_ref = tcx.impl_trait_ref(impl_def_id).unwrap().instantiate_identity();
     trait_ref.error_reported()?;
 
-    let ret = do_orphan_check_impl(tcx, trait_ref, impl_def_id);
-    if tcx.trait_is_auto(trait_ref.def_id) {
-        lint_auto_trait_impl(tcx, trait_ref, impl_def_id);
-    }
-
-    ret
+    do_orphan_check_impl(tcx, trait_ref, impl_def_id)
```
Comment (spastorino) [inline compiler/rustc_hir_analysis/src/coherence/orphan.rs:21]: I think it's probably better if we inline `do_orphan_check_impl`.

File: tests/ui/auto-traits/suspicious-negative-impls-lint.rs [hunk 4]
```diff
@@ -1,21 +1,16 @@
+//@ check-pass
```
Comment (lcnr) [inline tests/ui/auto-traits/suspicious-negative-impls-lint.rs:1]: these suspicious impls lint tests are now useless, please remove them entirely

**Comment by lcnr** [comment]
r=me after updating description and final nit. please also add closes #93367

**Pull request merged**